libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs

```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Plutus.Guardrail (guardrail)
+where
+
+import Cardano.Ledger.Binary.Plain (decodeFullFromHexText)
+import Cardano.Ledger.Plutus.Language (
+  Language (..),
+  Plutus (..),
+  PlutusBinary (..),
+ )
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Short as SBS (toShort)
+import qualified Data.Text as Text (Text)
+import GHC.Stack
+
+decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
+decodeHexPlutus = either (error . show) (Plutus . PlutusBinary . SBS.toShort) . decodeFullFromHexText @ByteString
```

Comment:
I think it would be better to move this function to `Test.Cardano.Ledger.Plutus` due to its generality.

Also, I don't think it is necessary to decode into a `ByteString` first. It should work equivalently well by decoding directly into a `ShortByteString`
```suggestion
decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
decodeHexPlutus = either (error . show) (Plutus . PlutusBinary) . decodeFullFromHexText @ByteString
```

By the way, the reason why decoding directly into `Plutus l` did not work for you, because that is not the decoder we use for dealing with scripts on mainnet:
https://github.com/IntersectMBO/cardano-ledger/blob/373c927b6fca77bd997624fd84567e280175a561/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs#L582-L584

Which has to do with the fact that tag is reused for native scripts and plutus scripts at the same time.

In other words decoder for `Plutus l` encodes the language as a tag (which is needed for encoding `PlutusWithContext` for debugging), while decoder for scripts on chain (encoded transactions) uses a different tag for this purpose. The script itself never actually encodes the plutus version, since binary representation of a plutus script could often be used for different plutus versions.

Reply:
Thanks for explanation! 
Moved the function to Plutus module and deserialized directly as ShortByteString as you suggested :ok_hand: